src/fs/file.rs

```diff
@@ -173,9 +173,13 @@ impl File {
     ///     })
     /// }
     /// ```
-    pub async fn read_at<T: IoBufMut>(&self, buf: T, pos: u64) -> crate::BufResult<usize, T> {
+    pub async fn read_at<T: IoBufMut>(
+        &self,
+        buf: impl Into<Slice<T>>,
+        pos: u64,
+    ) -> crate::BufResult<usize, Slice<T>> {
```

Comment:
To follow https://github.com/tokio-rs/tokio-uring/pull/163#discussion_r1013275409, perhaps the return type could be `BufResult<(), Slice<T>>`, and the returned slice's end is adjusted to reference the bytes that have been read. But that's a more extensive API change.

Reply:
Yes. If it hasn't been done, yet, now is probably not the time. This ownership API is probably new for anyone coming to this runtime and is going to take some getting used to.

Reply:
Another way of thinking about this (not in this PR), is when we do a read, should we read into the beginning of the buffer, or the first uninitialised byte?

Because if its into the first uninitialised byte, the returned usize still serves a purpose.

Reply:
I think the use of Slice as an input to the read op makes this easy since the definition can be to always read into the beginning of the slice. Let the caller define the slice to start at the uninitialized spot if they like. And I think that's what is currently done.

Reply:
The reading starts into the beginning of the buffer. This is consistent with the behaviour of `tokio::io` and `std::io`, where similar buffer adapters are used. Buffers are meant to be reused and overwritten, but a buffer should not need to be completely initialized (the innate problem of `Read::read`) in order to be used safely.

Reply:
Agree.